src/http.ts

```diff
@@ -0,0 +1,134 @@
+import fetch from 'unfetch';
+
+import {
+  DEFAULT_FETCH_TIMEOUT_MS,
+  DEFAULT_SILENT_TOKEN_RETRY_COUNT
+} from './constants';
+
+import { sendMessage } from './worker/worker.utils';
+import { FetchOptions } from './global';
+
+export const createAbortController = () => new AbortController();
+
+const switchFetch = async (
+  url: string,
+  audience: string,
+  scope: string,
+  opts: FetchOptions,
+  timeout: number,
+  worker?: Worker
+) => {
+  if (worker) {
+    // AbortSignal is not serializable, need to implement in the Web Worker
+    delete opts.signal;
+
+    return sendMessage(
+      {
+        auth: {
+          audience,
+          scope
+        },
+        timeout,
+        fetchUrl: url,
+        fetchOptions: { ...opts }
+      },
+      worker
+    );
+  } else {
+    const response = await fetch(url, opts);
+    return {
+      ok: response.ok,
+      json: await response.json()
+    };
+  }
+};
+
+export const fetchWithTimeout = (
+  url: string,
+  audience: string,
+  scope: string,
+  options: FetchOptions,
+  worker?: Worker,
+  timeout = DEFAULT_FETCH_TIMEOUT_MS
+) => {
+  const controller = createAbortController();
+  const signal = controller.signal;
+
+  const fetchOptions = {
+    ...options,
+    signal
+  };
+
+  let timeoutId: ReturnType<typeof setTimeout>;
```

Comment:
I had to google this, why are we using this? I know this was already in place, just asking...

Reply:
Not sure, could just as easily be `NodeJS.Timeout`. The way it's written here is a convenience of being able to just use the return type of whatever `setTimeout` returns, without knowing up-front what that type is.

Reply:
Not sure if it every different from number, but maybe it is ?